### PR TITLE
box: use 0 as a space id for FK that refer to the same space

### LINKED ISSUE #7200: Use "0" as a space_id or drop space_id for FK constraints that refer to themselves
Since we are using a space ID for FK constraints, it is inconvenient to create self-referencing FK constraints, as the space ID will only be generated during space creation. This is especially inconvenient for SQL, since the format for the space is created at compile time, and the space ID is only obtained at run time.

### PR DESCRIPTION
**Pull request by Gumix**
It is inconvenient to create self-referencing FK constraints, as the
space ID will only be generated during space creation. This is
especially useful for SQL, since the format for the space is created
at compile time, and the space ID is only obtained at run time.

Closes #7200

### CODE REVIEW

File: test/engine-luatest/gh_6961_fkey_same_space_test.lua
Comment (ImeevMA) [inline test/engine-luatest/gh_6961_fkey_same_space_test.lua:96]: I would add something like this:
```
tarantool> opts = setmetatable({}, { __serialize = 'map' }) 
---
...

tarantool> format = {{name = 'i', type = 'integer'}, {name = 'a', type = 'integer', foreign_key = {a1 = {space = 0, field = 'i'}}}}
---
...

tarantool> a = {512, 1, 'a', 'memtx', 0, opts, format}
---
...

tarantool> box.space._space:insert(a) 
---
- [512, 1, 'a', 'memtx', 0, {}, [{'name': 'i', 'type': 'integer'}, {'type': 'integer',
      'name': 'a', 'foreign_key': {'a1': {'space': 0, 'field': 'i'}}}]]
...

tarantool> box.space._space:get(512)
---
- [512, 1, 'a', 'memtx', 0, {}, [{'name': 'i', 'type': 'integer'}, {'type': 'integer',
      'name': 'a', 'foreign_key': {'a1': {'space': 0, 'field': 'i'}}}]]
...
```
I mean this problem is mainly for cases where you need to insert directly into _space (for example from SQL) so it makes sense to check if this works correctly. And, maybe, it makes sense to add similar tests for tuple FK.

Also, since this is labelled as a bug, I think it's better to create a separate file for this. Or at least replace the `bug` label with a `feature` label.

File: changelogs/unreleased/gh-7200-use-0-as-space-id-for-self-fkey.md
Comment (ImeevMA) [inline changelogs/unreleased/gh-7200-use-0-as-space-id-for-self-fkey.md:1]: Commit title is too long. It should be no longer than 50 symbols: https://github.com/tarantool/tarantool/wiki/Code-review-procedure#general-coding-points-to-check

**Review by ImeevMA — COMMENTED** [review]
Hi! Thank you for the review. I have two comments.

**Comment by ImeevMA** [comment]
Sorry, I have two more comments.

First: I believe we cannot say that this commit is refactoring in `NO_DOC` since it changes behavior.
Second: May be it makes sense to allow 0 in given format, since otherwise we will get problems like this:
```
tarantool> a = box.schema.space.create('a', {format = {{'i', 'integer'}, {name = 'a', type = 'integer', foreign_key = {a1 = {space = 'a', field = 'i'}}}}})
---
...

tarantool> a:format(a:format())
---
- error: 'Illegal parameters, format[2]: foreign key: space 0 was not found'
...
```

@alyapunov what do you think?

File: src/box/lua/schema.lua
Comment (Gumix) [reply src/box/lua/schema.lua:600]: @alyapunov?

Comment (alyapunov) [reply src/box/lua/schema.lua:600]: Sounds good, but it must be consistent with the Lua interface. I mean we must omit space_id in the Lua interface too. We should discuss it with the issue owner.

**Comment by alyapunov** [comment]
> `NO_DOC`

I agree, we have to make changes in doc too.

And a:format(a:format()) must work, but perhaps with #7350.

**Review by alyapunov — CHANGES_REQUESTED** [review]

File: test/engine-luatest/gh_6961_fkey_same_space_test.lua
Comment (Gumix) [reply test/engine-luatest/gh_6961_fkey_same_space_test.lua]: Done

File: src/box/tuple_constraint_fkey.c
Comment (Gumix) [reply src/box/tuple_constraint_fkey.c]: Done

File: src/box/lua/schema.lua
Comment (Gumix) [reply src/box/lua/schema.lua]: Done

**Comment by Gumix** [comment]
> First: I believe we cannot say that this commit is refactoring in `NO_DOC` since it changes behavior.
Done.

> Second: May be it makes sense to allow 0 in given format, since otherwise we will get problems like this:
Fixed by omitting the `space` field.

**Review by locker — APPROVED** [review]

**Review by ImeevMA — APPROVED** [review]
Thanks for the patch!

**Comment by locker** [comment]
Can't merge until @alyapunov approves :-/

image

**Review by alyapunov — APPROVED** [review]

**Pull request merged**
